### PR TITLE
[BottomNavigation] Adds state restoration support to the bottom navigation bar controller.

### LINKED ISSUE #8265: [BottomNavigation] Support state restoration
Clients who use state restoration in their application should be able to use `MDCBottomNavigationBarController` as part of saving/restoring their application state.

### PR DESCRIPTION
**Pull request by nokeeo**
Enables clients to opt-in to state restoration by providing `restorationIdentifier` values for the Bottom Navigation controller's child view controllers.

Closes #8265

### CODE REVIEW

File: components/BottomNavigation/src/MDCBottomNavigationBarController.m [hunk 1]
```diff
@@ -526,7 +571,7 @@ - (void)showLargeItemDialog {
   [self.largeItemDialog.centerYAnchor constraintEqualToAnchor:window.centerYAnchor].active = YES;
 
   self.largeItemDialog.layer.opacity = 0;
-  self.largeItemDialog.transform = MDCLargeItemViewAnimationTransitionTransform();
+  self.largeItemDialog.transform = LargeItemViewAnimationTransitionTransform();
```
Comment (romoore) [inline components/BottomNavigation/src/MDCBottomNavigationBarController.m]: Nit: Let's move this refactoring to a separate PR in case we need to roll-back the state restoration change.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m [hunk 2]
```diff
@@ -313,7 +368,56 @@ - (void)testNonSelectedViewControllersHaveViewsLazilyLoaded {
   XCTAssertFalse(childViewController2.viewLoaded);
 }
 
-#pragma mark - MDCBottomNavigationBarControllerDelegate Methods
+- (void)testEncodesChildViewControllersWhenPreservingRestorationState {
+  // Given
+  UIViewController *childViewController1 = [[UIViewController alloc] init];
+  childViewController1.restorationIdentifier = @"vc1";
```
Comment (romoore) [inline components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m:374]: Can you have each VC retain some unique properties? Perhaps something simple like `title` so it can be verified we didn't mix them up?

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m [hunk 3]
```diff
@@ -338,9 +442,9 @@ - (BOOL)bottomNavigationBarController:
   return [self.delegateReturnValue boolValue];
 }
 
-#pragma mark - Helper Methods
+#pragma mark Helper Methods
```
Comment (romoore) [inline components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m]: Nit: I would prefer to keep the divider.  In either case, please separate the refactoring/clean-up to a different PR.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m [hunk 4]
```diff
@@ -313,7 +368,56 @@ - (void)testNonSelectedViewControllersHaveViewsLazilyLoaded {
...
+  UIViewController *childViewController1 = [[UIViewController alloc] init];
+  childViewController1.restorationIdentifier = @"vc1";
+
+  UIViewController *childViewController2 = [[UIViewController alloc] init];
+  childViewController2.restorationIdentifier = @"vc2";
+
+  self.bottomNavigationBarController.viewControllers =
+      @[ childViewController1, childViewController2 ];
+  self.bottomNavigationBarController.viewControllers[0].restorationIdentifier = @"vc1";
+  self.bottomNavigationBarController.viewControllers[1].restorationIdentifier = @"vc2";
```
Comment (romoore) [inline components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m]: Is this duplicating the property assignments above or am I missing something?

**Review by romoore — APPROVED** [review]
Thanks for the contribution! LGTM after moving the clean-up/refactoring to a different PR.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m (hunk 4 above)
Comment (nokeeo) [reply components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m]: Done.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m (hunk 2 above)
Comment (nokeeo) [reply components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m:374]: Sorry not following.  This tests that the archive simply contains the encoded view controller for a restoration identifier.  The mapping of restoration identifier to view controller is normally handled by UIKit.  I think it might be useful to assert that a property remains constant when we decode the view controllers in the other test?

File: components/BottomNavigation/src/MDCBottomNavigationBarController.m (hunk 1 above)
Comment (nokeeo) [reply components/BottomNavigation/src/MDCBottomNavigationBarController.m]: Done.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m (hunk 3 above)
Comment (nokeeo) [reply components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m]: Done.

File: components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m (hunk 2 above)
Comment (romoore) [reply components/BottomNavigation/tests/unit/MDCBottomNavigationBarControllerTests.m:374]: Ah, ok. That makes more sense. I probably misinterpreted the name of the test to indicate that it correctly encodes the child VCs, not that the coder contains the child VCs.

**Review by romoore — APPROVED** [review]

File: components/BottomNavigation/src/MDCBottomNavigationBarController.m [hunk 5]
```diff
@@ -27,6 +27,7 @@
 static const NSTimeInterval kLargeItemViewAnimationDuration = 0.1;
 static const NSTimeInterval kLongPressMinimumPressDuration = 0.2;
 static const NSUInteger kLongPressNumberOfTouchesRequired = 1;
+static NSString *const kSelectedViewControllerRestorationKey = @"selectedViewController";
```
Comment (romoore) [inline components/BottomNavigation/src/MDCBottomNavigationBarController.m:30]: Should this key be unique to this class? Maybe something like `MDCBottomNavigationBarControllerSelectedViewControllerKey`?

Comment (nokeeo) [reply components/BottomNavigation/src/MDCBottomNavigationBarController.m:30]: The given coder is scoped to the particular controller.  The WWDC presentation introducing state restoration says you do not have to do this.

Comment (romoore) [reply components/BottomNavigation/src/MDCBottomNavigationBarController.m:30]: Awesome, thanks for clarifying!

**Comment by romoore** [comment]
@nokeeo I will be OOO Monday and Tuesday. Could you resolve the conflicts so this can be merged?

**Comment by nokeeo** [comment]
> @nokeeo I will be OOO Monday and Tuesday. Could you resolve the conflicts so this can be merged?

@romoore Ok no rush on this!  Resolved conflicts.
